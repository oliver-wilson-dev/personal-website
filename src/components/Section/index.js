import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import sharedStyles from '../App/index.css';
import styles from './index.css';
import useFadeInClasses from '../../hooks/useFadeInClasses';

const Section = ({ title, children, additionalStyles = {} }) => {
  const { fadeInClasses } = useFadeInClasses();

  return (
    <section className={cn(
      styles.flexColumn,
      sharedStyles.flexCenter,
      fadeInClasses,
      { [additionalStyles.section]: !!additionalStyles.section, }
    )}
    >
      {title && <h2 className={styles.sectionTitle}>{title}</h2>}
      <div className={additionalStyles.container}>
        {children}
      </div>
    </section>
  );
};

Section.propTypes = {
  title: PropTypes.string,
  children: PropTypes.node.isRequired,
  additionalStyles: PropTypes.oneOfType([
    PropTypes.shape()
  ])
};

export default Section;
