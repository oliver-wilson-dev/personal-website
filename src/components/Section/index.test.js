import React from 'react';
import { mount } from 'enzyme';
import Section from '.';
import useFadeInClasses from '../../hooks/useFadeInClasses';

jest.mock('../SectionTile', () => {
  const SectionTile = () => null;

  return SectionTile;
});

jest.mock('../../hooks/useFadeInClasses');

const defaultProps = {
  title: 'test-title',
  children: [<p key="test-key">this is some content</p>]
};

const render = ({ overrideProps } = { overrideProps: {} }) => mount(
  <Section
    {...defaultProps}
    {...overrideProps}
  />
);

describe('<Section/>', () => {
  beforeEach(() => {
    useFadeInClasses.mockReturnValue({ fadeInClasses: 'test-fade-in-classes' });
  });


  describe('when the children are HTML elements', () => {
    it('should exist', () => {
      expect(render().exists()).toBe(true);
    });

    it('should render correctly', () => {
      expect(render()).toMatchSnapshot();
    });
  });

  describe('when children are provided', () => {
    const htmlElement = <p key="test-key">some component</p>;
    const SomeComponent = () => htmlElement;

    it('should exist', () => {
      expect(render({
        overrideProps: {
          children: [<SomeComponent key="1" />, htmlElement]
        }
      }).exists()).toBe(true);
    });

    it('should render correctly', () => {
      expect(render({
        overrideProps: {
          children: [<SomeComponent key="1" />, htmlElement]
        }
      })).toMatchSnapshot();
    });
  });

  describe('when there is no title', () => {
    it('should render correctly', () => {
      expect(render({ title: undefined }).exists()).toBe(true);
    });
  });

  describe('when additionalStyles are provided', () => {
    describe('when a section class is provided', () => {
      it('should add the class to the section element', () => {
        const sectionClass = 'test-section-class';
        const wrapper = render({
          overrideProps: {
            additionalStyles: {
              section: sectionClass
            }
          }
        });

        expect(wrapper.find('section').hasClass(sectionClass)).toBe(true);
      });
    });

    describe('when a container class is provided', () => {
      it('should add the class to the section element', () => {
        const containerClass = 'test-container-class';
        const wrapper = render({
          overrideProps: {
            additionalStyles: {
              container: containerClass
            }
          }
        });

        expect(wrapper.find(`.${containerClass}`).exists()).toBe(true);
      });
    });
  });
});
