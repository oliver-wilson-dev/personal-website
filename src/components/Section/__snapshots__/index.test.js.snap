// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Section/> when children are provided should render correctly 1`] = `
<Section
  title="test-title"
>
  <section
    className="flexColumn flexCenter test-fade-in-classes"
  >
    <h2
      className="sectionTitle"
    >
      test-title
    </h2>
    <SomeComponent
      key="1"
    >
      <p
        key="test-key"
      >
        some component
      </p>
    </SomeComponent>
    <p
      key="test-key"
    >
      some component
    </p>
  </section>
</Section>
`;

exports[`<Section/> when the children are HTML elements should render correctly 1`] = `
<Section
  title="test-title"
>
  <section
    className="flexColumn flexCenter test-fade-in-classes"
  >
    <h2
      className="sectionTitle"
    >
      test-title
    </h2>
    <p
      key="test-key"
    >
      this is some content
    </p>
  </section>
</Section>
`;
